### PR TITLE
Improved sliding window - trim old context

### PR DESCRIPTION
**Pull request by conceptd3v**
and also trim 'current' context if it exceeds the limit (rare but possible bug)

### CODE REVIEW

File: Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt [hunk 1]
```diff
@@ -36,9 +57,30 @@ class CustomChatMessage @OptIn(BetaOpenAI::class) constructor(
         return messageTokens
     }
 
+    /**
+     *
+     */
```
Comment (khalp) [inline Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt]: can remove or add header comment !

File: Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt [hunk 2]
```diff
@@ -36,9 +57,30 @@ class CustomChatMessage @OptIn(BetaOpenAI::class) constructor(
...
+    public fun canFitInTokenLimit(includeGrounding: Boolean = true, tokensAllowed: Int = -1): Boolean {
+        if (tokensAllowed < 0) return true
+        if (getTokenCount(includeGrounding, tokensAllowed) <= tokensAllowed) return true
+        return false
```
Comment (khalp) [inline Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt]:
```suggestion
        return getTokenCount(includeGrounding, tokensAllowed) <= tokensAllowed
```

File: Jetchat/app/src/main/java/com/example/compose/jetchat/data/Tokenizer.kt [hunk 3]
```diff
@@ -22,5 +22,21 @@ class Tokenizer {
...
+        /**
+         * Trim the input text to be under the number of tokens specified
+         *
+         * @return substring of the text that's no longer than the
+         * specified number of tokens
+         */
+        fun trimToTokenLimit (text: String?, tokenLimit: Int): String? {
+            // TODO: limit by tokens instead of the rough character approximation
+            val charLimit = tokenLimit * 4
+            return if (text?.length!! <= charLimit) {
```
Comment (khalp) [inline Jetchat/app/src/main/java/com/example/compose/jetchat/data/Tokenizer.kt]:
```suggestion
            val length = text?.length
            return if (length != null && length <= charLimit) {
```

Update logic to allow for possibly null text field (`!!` assertion will throw exception)

File: Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt [hunk 4]
```diff
@@ -36,9 +57,30 @@ class CustomChatMessage @OptIn(BetaOpenAI::class) constructor(
...
+    }
+
+    /**
+     * Create `ChatMessage` instance to add to completion request
+     */
     @OptIn(BetaOpenAI::class)
-    public fun getChatMessage () : ChatMessage {
-        val content = grounding + userContent
+    public fun getChatMessage (includeGrounding: Boolean = true, tokensAllowed: Int = -1) : ChatMessage {
```
Comment (khalp) [inline Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt]:
```suggestion
    fun getChatMessage (includeGrounding: Boolean = true, tokensAllowed: Int = -1) : ChatMessage {
```

File: Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt [hunk 5]
```diff
@@ -36,9 +57,30 @@ class CustomChatMessage @OptIn(BetaOpenAI::class) constructor(
         return messageTokens
     }
 
+    /**
+     *
+     */
+    public fun canFitInTokenLimit(includeGrounding: Boolean = true, tokensAllowed: Int = -1): Boolean {
```
Comment (khalp) [inline Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt]:
```suggestion
    fun canFitInTokenLimit(includeGrounding: Boolean = true, tokensAllowed: Int = -1): Boolean {
```

File: Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt [hunk 6]
```diff
@@ -26,8 +29,26 @@ class CustomChatMessage @OptIn(BetaOpenAI::class) constructor(
...
+     * be a validation error on the UI)
+     */
+    public fun getTokenCount(includeGrounding: Boolean = true, tokensAllowed: Int = -1) : Int {
+        var messageContent = userContent ?: ""
+        if (includeGrounding) {
+            messageContent = if (tokensAllowed < 0) {
+                grounding + userContent ?: ""
+            } else { // only include as much of the grounding as will fit
+                Tokenizer.trimToTokenLimit(grounding, tokensAllowed) + userContent ?: ""
+            }
```
Comment (khalp) [inline Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt]:
```suggestion
        if (includeGrounding) {
            messageContent = if (tokensAllowed < 0) {
                grounding + messageContent
            } else { // only include as much of the grounding as will fit
                Tokenizer.trimToTokenLimit(grounding, tokensAllowed) + messageContent
            }
```

File: Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt [hunk 7]
```diff
@@ -26,8 +29,26 @@ class CustomChatMessage @OptIn(BetaOpenAI::class) constructor(
...
+     * Count the number of tokens in the user query PLUS the additional
+     * embedding data for grounding OR the functions when userContent is empty
+     *
+     * `userContent` can never be larger than `tokensAllowed` - if this happens in
+     * message history, the message will be dropped from the window. In theory
+     * it could happen if a user entered maxTokens worth of text in their
+     * chat query, but in practice that seems unlikely (and should probably
+     * be a validation error on the UI)
+     */
+    public fun getTokenCount(includeGrounding: Boolean = true, tokensAllowed: Int = -1) : Int {
```
Comment (khalp) [inline Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt]:
```suggestion
    fun getTokenCount(includeGrounding: Boolean = true, tokensAllowed: Int = -1) : Int {
```

File: Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt [hunk 8]
```diff
@@ -5,14 +5,17 @@ import com.aallam.openai.api.chat.ChatMessage
...
-    val role: ChatRole,
-    val grounding: String? = null,
-    val userContent: String? = null,
-    val name: String? = null,
-    val functionCall: FunctionCall? = null
+    public val role: ChatRole,
+    public val grounding: String? = null,
+    public val userContent: String? = null,
+    public val name: String? = null,
+    public val functionCall: FunctionCall? = null
```
Comment (khalp) [inline Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt]:
```suggestion
    val role: ChatRole,
    val grounding: String? = null,
    val userContent: String? = null,
    val name: String? = null,
    val functionCall: FunctionCall? = null
```

I think the IDE is suggesting that the public modifier is unnecessary - but if you needed them to be added feel free to disregard!

File: Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt [hunk 9]
```diff
@@ -5,14 +5,17 @@ import com.aallam.openai.api.chat.ChatMessage
...
     ) {
 
     public fun summary(): String? {
```
Comment (khalp) [inline Jetchat/app/src/main/java/com/example/compose/jetchat/data/CustomChatMessage.kt]:
```suggestion
    fun summary(): String? {
```

File: Jetchat/app/src/main/java/com/example/compose/jetchat/DroidconEmbeddingsWrapper.kt [hunk 10]
```diff
@@ -296,17 +299,20 @@ class DroidconEmbeddingsWrapper(val context: Context?) {
...
-        var date = Constants.TEST_DATE
-        var time = Constants.TEST_TIME
-        if (date == "") {
-            date = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))
-        }
-        if (time == "") {
-            time  = LocalDateTime.now().format(DateTimeFormatter.ofPattern("HH:mm"))
+        if (messagePreamble.isNullOrEmpty()) {
+            // ONLY show date/time when embeddings are empty, as it triggers the SessionsByTime function (I THINK)
+            // ALWAYS add the date and time to every prompt
```
Comment (khalp) [inline Jetchat/app/src/main/java/com/example/compose/jetchat/DroidconEmbeddingsWrapper.kt:304]:
```suggestion
            // ONLY show date/time when embeddings are empty, as it triggers the SessionsByTime function
```

hoping the "i think" has been resolved 😜

File: Jetchat/app/src/main/java/com/example/compose/jetchat/DroidconEmbeddingsWrapper.kt [hunk 11]
```diff
@@ -196,19 +199,19 @@ class DroidconEmbeddingsWrapper(val context: Context?) {
...
             }
             if (handled) {
                 // add the 'call a function' response to the history
                 conversation.add(
-                    ChatMessage(
+                    CustomChatMessage(
                         role = completionMessage.role,
-                        content = completionMessage.content
+                        userContent = completionMessage.content
                             ?: "", // required to not be empty in this case
```
Comment (khalp) [inline Jetchat/app/src/main/java/com/example/compose/jetchat/DroidconEmbeddingsWrapper.kt:215]: is this still true with the custom chat message class? or can we remove the null check now

**Review by khalp — APPROVED** [review]
looks great! added nitpicky suggestions, feel free to reject

File: Jetchat/app/src/main/java/com/example/compose/jetchat/DroidconEmbeddingsWrapper.kt (hunk 10 above)
Comment (conceptdev) [reply Jetchat/app/src/main/java/com/example/compose/jetchat/DroidconEmbeddingsWrapper.kt:304]: 🤔 not sure!

**Pull request merged**